File: src/main/java/schemabuilder/processor/pipelines/parsing/directives/GraphQLDirectiveParser.java
package schemabuilder.processor.pipelines.parsing.directives;

import graphql.schema.idl.SchemaDirectiveWiring;
import schemabuilder.annotations.graphql.GraphQLDirective;
import schemabuilder.processor.pipelines.parsing.GraphQLClassParserStrategy;
import schemabuilder.processor.wiring.InstanceFetcher;

public class GraphQLDirectiveParser implements GraphQLClassParserStrategy {

    private GraphQLDirectiveBank directiveBank = GraphQLDirectiveBank.getInstance();

    @Override
    public void parse(Class<?> clazz, InstanceFetcher fetcher) {
        if (!clazz.isAnnotationPresent(GraphQLDirective.class)) {
            return;
        }

        String typeName = clazz.getAnnotation(GraphQLDirective.class).value();
        Object instance = fetcher.getInstance(clazz);

        if (!(instance instanceof SchemaDirectiveWiring)) {
            return;
        }

        directiveBank.addDirective(new GraphQLDirectiveType(typeName, (SchemaDirectiveWiring) instance));
    }
}

File: src/main/java/schemabuilder/processor/pipelines/parsing/directives/GraphQLDirectiveParser.kt
package schemabuilder.processor.pipelines.parsing.directives

import graphql.schema.idl.SchemaDirectiveWiring
import schemabuilder.annotations.graphql.GraphQLDirective
import schemabuilder.processor.pipelines.parsing.GraphQLClassParserStrategy
import schemabuilder.processor.wiring.InstanceFetcher

class GraphQLDirectiveParser : GraphQLClassParserStrategy {

    override fun parse(clazz: Class<*>, fetcher: InstanceFetcher) {
        val typeName: String = clazz.getAnnotation(GraphQLDirective::class.java).value
        val instance = fetcher.getInstance(clazz) as? SchemaDirectiveWiring ?: return
        GraphQLDirectiveBank.addDirective(GraphQLDirectiveType(typeName, instance))
    }

    override fun shouldParse(clazz: Class<*>): Boolean = clazz.isAnnotationPresent(GraphQLDirective::class.java)
}